### ISSUE TITLE
Is the leveldb background compact single threaded?

### ISSUE THREAD

**Issue by joscul** [google/leveldb#977]
First of all, this is a great library. But after using it a bit making huge amounts of inserts I have two questions.

1. Is the leveldb background compact job single threaded?
2. Is it possible to do inserts while the background compact job is running or is the database locked?

I am using WriteBatch to make 1M inserts at a time. At regular intervals the inserts are just waiting for around 300 seconds before continuing. My guess is that the compact job is running and taking up only one cpu.

**Comment by nerdondon** [comment]
Having a read through a large portion of the code base, I feel confident to say (correct me if I'm wrong):

1. The background compaction job is single threaded
2. It is possible to do inserts while the background compaction job is running, however certain situations will cause the writing thread to wait. Namely, when a memtable is full and a previous memtable is already undergoing compaction. This will pause writing threads until a new memtable can be created i.e. when the compaction is done. The default memtable size is 4 MiB. Maybe your write pattern is filling memtables up too fast?

Aside, these types of questions are probably best suited to be asked on stack overflow. From what I've seen, unless its an actual problem with the lib, these types of questions dont really get much love in the leveldb issues.

**Comment by joscul** [comment]
Thanks for the hint.

I also started reading the source and the background worker seems to be very single threaded. Even if I create multiple instances (pointing to separate db folders) the background worker is shared between them. It runs as a singleton.
https://github.com/google/leveldb/blob/4fb146810cd265ffefa7f9905c016ae965ad36c8/util/env_posix.cc#L922

I ended up increasing the write_buffer_size to 1GB.
```
options.write_buffer_size = 1024ull * 1024ull * 1024ull
```
This made the problem less frequent anyway.

Ok, will post on stack overflow next time. Closing this thread.
